spark-statsd-reporter/src/main/java/org/apache/spark/metrics/sink/statsd/InstanceDetailsProvider.java

```diff
@@ -18,23 +18,29 @@
     private final static Logger logger = LoggerFactory.getLogger(StatsdReporter.class);
 
     private InstanceDetails instance = null;
+    private boolean initialized = false;
 
     Optional<InstanceDetails> getInstanceDetails() {
-        if(instance == null) {
+        if (!initialized) {
             if (SparkEnv.get() == null) {
                 logger.warn("SparkEnv is not initialized, instance details unavailable");
             } else {
-                SparkConf sparkConf = SparkEnv.get().conf();
-                instance =
-                        new InstanceDetails(
-                                sparkConf.getAppId(),
-                                sparkConf.get("spark.app.name"),
-                                InstanceType.valueOf(SparkEnv.get().metricsSystem().instance().toUpperCase()),
-                                sparkConf.get("spark.executor.id"),
-                                sparkConf.get("spark.metrics.namespace", "default")
-                        );
+                instance = buildInstanceDetails();
             }
+            initialized = true;
```

Comment:
if the instance is null the first time, it's okay that it remain null for its lifetime? You never want to `buildInstanceDetails()` later ? I guess that would mean `SparkEnv` is never changed?

Reply:
This extra-check was introduced for lazy evaluation purposes and to avoid noisy logging every time it is `null`. The method returns `Optional.ofNullable(instance)` to indicate a possible absence of `SparkEnv` so it's fine for the instance to stay `null` (this use case is Dispatcher-specific, Driver and Executors always have `SparkEnv` initialized).

Reply:
ah thanks for the explanation!